substrate/frame/support/procedural/tools/src/lib.rs

```diff
@@ -39,25 +39,47 @@ fn generate_hidden_includes_mod_name(unique_id: &str) -> Ident {
 /// Generates the access to the `frame-support` crate.
 pub fn generate_crate_access(unique_id: &str, def_crate: &str) -> TokenStream {
 	if std::env::var("CARGO_PKG_NAME").unwrap() == def_crate {
-		quote::quote!(frame_support)
+		let frame_support = match generate_crate_access_2018("frame-support") {
+			Ok(c) => c,
+			Err(e) => return e.into_compile_error().into(),
+		};
+		quote::quote!(#frame_support)
 	} else {
 		let mod_name = generate_hidden_includes_mod_name(unique_id);
 		quote::quote!( self::#mod_name::hidden_include )
 	}
 }
 
+/// Check if a path is using the `frame` crate or not.
+///
+/// This will usually check the output of [`generate_crate_access_2018`].
+/// We want to know if whatever the `path` takes us to, is exported from `frame` or not. In that
+/// case `path` would start with `frame`, something like `frame::x::y:z`.
+pub fn is_using_frame_crate(path: &syn::Path) -> bool {
+	path.segments.first().map(|s| s.ident == "frame").unwrap_or(false)
+}
+
 /// Generate the crate access for the crate using 2018 syntax.
 ///
-/// for `frame-support` output will for example be `frame_support`.
-pub fn generate_crate_access_2018(def_crate: &str) -> Result<syn::Ident, Error> {
-	match crate_name(def_crate) {
+/// If `frame` is in scope, it will use `frame::deps::<def_crate>`. Else, it will try and find
+/// `<def_crate>` directly.
+pub fn generate_crate_access_2018(def_crate: &str) -> Result<syn::Path, Error> {
```

Comment:
In other crates we might have the same function but which doesn't generate access from `frame`.

Maybe it worth giving a different name like `generate_crate_access_from_frame_or_deps`.

(But this is minor, it is also ok)